src/zcash/circuit/gadget.tcc

```diff
@@ -0,0 +1,349 @@
+                right_side
+            ));
+
+            // #854: Ensure that left_side is a 64-bit integer.
+            for (size_t i = 0; i < 64; i++) {
+                generate_boolean_r1cs_constraint<FieldT>(
+                    this->pb,
+                    zk_total_uint64[i],
+                    ""
+                );
+            }
+
+            this->pb.add_r1cs_constraint(r1cs_constraint<FieldT>(
+                1,
+                left_side,
+                packed_addition(zk_total_uint64)
+            ));
+        }
+    }
+
+    void generate_r1cs_witness(
+        const uint256& phi,
+        const uint256& rt,
+        const uint256& h_sig,
+        const boost::array<JSInput, NumInputs>& inputs,
+        const boost::array<Note, NumOutputs>& outputs,
+        uint64_t vpub_old,
+        uint64_t vpub_new
+    ) {
+        // Witness `zero`
+        this->pb.val(ZERO) = FieldT::zero();
+
+        // Witness rt. This is not a sanity check.
+        //
+        // This ensures the read gadget constrains
+        // the intended root in the event that
+        // both inputs are zero-valued.
+        zk_merkle_root->bits.fill_with_bits(
+            this->pb,
+            uint256_to_bool_vector(rt)
+        );
+
+        // Witness public balance values
+        zk_vpub_old.fill_with_bits(
+            this->pb,
+            uint64_to_bool_vector(vpub_old)
+        );
+        zk_vpub_new.fill_with_bits(
+            this->pb,
+            uint64_to_bool_vector(vpub_new)
+        );
+
+        {
+            // Witness total_uint64 bits
+            uint64_t left_side_acc = vpub_old;
+            for (size_t i = 0; i < NumInputs; i++) {
+                left_side_acc += inputs[i].note.value;
+            }
+
+            zk_total_uint64.fill_with_bits(
+                this->pb,
+                uint64_to_bool_vector(left_side_acc)
+            );
+        }
+
+        // Witness phi
+        zk_phi->bits.fill_with_bits(
+            this->pb,
+            trailing252(uint256_to_bool_vector(phi))
+        );
+
+        // Witness h_sig
+        zk_h_sig->bits.fill_with_bits(
+            this->pb,
+            uint256_to_bool_vector(h_sig)
+        );
+
+        for (size_t i = 0; i < NumInputs; i++) {
+            // Witness the input information.
+            auto merkle_path = inputs[i].witness.path();
+            zk_input_notes[i]->generate_r1cs_witness(
+                merkle_path,
+                inputs[i].key,
+                inputs[i].note
+            );
+
+            // Witness macs
+            zk_mac_authentication[i]->generate_r1cs_witness();
+        }
+
+        for (size_t i = 0; i < NumOutputs; i++) {
+            // Witness the output information.
+            zk_output_notes[i]->generate_r1cs_witness(outputs[i]);
+        }
+
+        // [SANITY CHECK] Ensure that the intended root
+        // was witnessed by the inputs, even if the read
+        // gadget overwrote it. This allows the prover to
```

Comment:
I don't understand "even if the read gadget overwrote it". How could that happen?